File: Hangman Equivalency/Hangman.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Text.RegularExpressions;
using System.Threading;
using System.Runtime.InteropServices.ComTypes;

namespace A01___Hangman
{
   class Hangman
   {
      readonly static Regex alphaAndSpaceReg = new Regex(@"^[a-zA-Z ]+$");
      readonly static Regex alphaCharReg = new Regex(@"[a-zA-Z]");
      readonly static Regex affirmativeReg = new Regex(@"^[Y|yE|eS|s]|[Y|y]");
      readonly static Regex negativeReg = new Regex(@"^[N|nO|o]|[N|n]|[Q|qU|uI|iT|t]");
      static void Main(string[] args)
      {
         //int[,] values = new int[4, 5]
         //    {
         //        { 1, 2, 3, 4, 5 },
         //        { 6, 7, 8, 9, 10 },
         //        { 11, 12, 13, 14, 15 },
         //        { 16, 17, 18, 19, 20 },
         //    };
         //for (int i = 0, j = 0; !(i == values.GetLength(0)); i += j == values.GetLength(1) - 1 ? 1 : 0, j += j < values.GetLength(1) - 1 ? 1 : -j)
         //{
         //    Console.Write(values[i, j] + " ");
         //}
         //Console.ReadLine();
         //return;
         bool running = true;
         while (running)
         {
            Console.Clear();
            bool gaming = true;
            bool validInput = false;
            string fullWord = string.Empty;
            StringBuilder guessWord = new StringBuilder(string.Empty);
            List<char> guessedLetters = new List<char>();
            int livesLeft = 6;
            Console.WriteLine("Welcome to hangman! Enter a word to start: (Can contain only letters and spaces.) ");
            while (!validInput)
            {
               // Cleans the input.
               fullWord = string.Join(" ", Console.ReadLine().Trim().Split(new char[] { ' ' }, StringSplitOptions.RemoveEmptyEntries)).ToLower();
               if (alphaAndSpaceReg.IsMatch(fullWord))
               {
                  validInput = true;
               }
               else
               {
                  Console.WriteLine("Sorry, that input isn't valid. You can only choose a word or phrase that contains letters and spaces.");
                  Console.WriteLine("Enter a word to start: (Can contain only letters and spaces.) ");
               }
            }
            // We now have a valid input.            
            guessWord.Append(alphaCharReg.Replace(fullWord, "_"));
            Console.Clear();
            //print status bar:
            Console.WriteLine($"Lives Left: {livesLeft}.   Letters guessed: {string.Join(", ", guessedLetters)}\n");
            //print word:
            Console.WriteLine(guessWord.ToString());
            while (gaming)
            {
               //print prompt:
               bool validGuess = false;
               char guess = '\0';
               while (!validGuess)
               {
                  Console.WriteLine("\nGuess a letter: ");
                  guess = Console.ReadLine().ToLower().Trim().FirstOrDefault();
                  if (guess == '\0')
                  {
                     Console.WriteLine("Make sure to type a character. The first character entered will be your guess.");
                  }
                  else if (guessedLetters.Contains(guess))
                  {
                     Console.WriteLine("You already guessed that letter!");
                  }
                  else if (!alphaCharReg.IsMatch(guess.ToString()))
                  {
                     Console.WriteLine("You can only guess English letters!");
                  }
                  else
                  {
                     validGuess = true;
                  }
               }
               guessedLetters.Add(guess);
               //evaluate:
               bool didFindLetter = false;
               for (int i = 0; i < fullWord.Length; i++)
               {
                  if (fullWord.ToLower()[i] == guess)
                  {
                     didFindLetter = true;
                     guessWord.Remove(i, 1);
                     if (i == 0 || fullWord[i - 1] == ' ')
                     {
                        guessWord.Insert(i, guess.ToString().ToUpper());
                     }
                     else
                     {
                        guessWord.Insert(i, guess.ToString().ToLower());
                     }
                  }
               }
               if (!didFindLetter)
               {
                  livesLeft--;
               }
               //Clear for fresh start.
               Console.Clear();
               //print status bar:
               Console.WriteLine($"Lives Left: {livesLeft}.   Letters guessed: {string.Join(", ", guessedLetters)}\n");
               //print word:
               Console.WriteLine(guessWord.ToString());
               //check for win/lose.
               if (livesLeft == 0 || !guessWord.ToString().Contains('_'))
               {
                  if (livesLeft == 0)
                  {
                     //You lose.
                     Console.WriteLine($"Uh oh! Out of lives, you lose. The word/phrase was: \"{fullWord}\" Play Again? (Y/N)");
                  }
                  else
                  {
                     //You win!
                     Console.WriteLine("Congratulations! You win! Play Again? (Y/N)");
                  }
                  string input = string.Empty;
                  validInput = false;
                  while (!validInput)
                  {
                     input = Console.ReadLine();
                     if (affirmativeReg.IsMatch(input) || negativeReg.IsMatch(input))
                     {
                        validInput = true;
                     }
                     else
                     {
                        Console.WriteLine("Hey! That isn't a valid input, say \"Y\" or \"N\".");
                     }
                  }
                  if (affirmativeReg.IsMatch(input))
                  {
                     livesLeft = 6;
                     guessedLetters.Clear();
                     guessWord.Clear();
                     fullWord = string.Empty;
                     gaming = false;
                     running = true;
                  }
                  else if (negativeReg.IsMatch(input))
                  {
                     Console.WriteLine("Thanks for playing!");
                     Thread.Sleep(2000);
                     gaming = false;
                     running = false;
                  }
               }
            }
         }
      }
   }
}

File: Hangman Equivalency/Hangman.fs
//This is a direct conversion of my C# version of this game.

open System
open System.Text.RegularExpressions
open System.Text
open System.Linq
open System.Collections.Generic

let alphaAndSpaceReg : Regex = new Regex(@"^[a-zA-Z ]+$")
let alphaCharReg : Regex = new Regex(@"[a-zA-Z]")
let affirmativeReg : Regex = new Regex(@"^[Y|yE|eS|s]|[Y|y]")
let negativeReg : Regex = new Regex(@"^[N|nO|o]|[N|n]|[Q|qU|uI|iT|t]")

[<EntryPoint>]
let main (argv : string[]) =
    let mutable running : bool = true
    while (running) do
        Console.Clear() |> ignore
        let mutable gaming : bool = true
        let mutable validInput : bool = false
        let mutable fullWord : string = String.Empty
        let guessWord : StringBuilder = new StringBuilder()
        let guessedLetters : List<char> = new List<char>()
        let mutable livesLeft : int = 6
        Console.WriteLine "Welcome to hangman! Enter a word to start: (Can contain only letters and spaces.)"
        while (not(validInput)) do
            //Cleans the input.
            fullWord <- String.Join(" ", Console.ReadLine().Trim().Split(' ', StringSplitOptions.RemoveEmptyEntries)).ToLower()
            if (alphaAndSpaceReg.IsMatch fullWord) then
                validInput <- true
            else
                Console.WriteLine("Sorry, that input isn't valid. You can only choose a word or phrase that contains letters and spaces.")
                Console.WriteLine("Enter a word to start: (Can contain only letters and spaces.) ")
        //We now have valid input.
        alphaCharReg.Replace(fullWord, "_") |> guessWord.Append |> ignore
        //Fresh start for printing UI.
        Console.Clear() |> ignore
        //Print status bar:
        Console.WriteLine("Lives Left: {0}.   Letters guessed: {1}\n", livesLeft, String.Join(", ", guessedLetters))
        //Print word:
        Console.WriteLine(guessWord.ToString())
        while (gaming) do
            //Print prompt:
            let mutable validGuess : bool = false
            let mutable guess : char = Char.MinValue
            while (not(validGuess)) do
                Console.WriteLine("\nGuess a letter: ")
                guess <- Console.ReadLine().ToLower().Trim().FirstOrDefault()
                if (guess = Char.MinValue) then
                    Console.WriteLine("Make sure to type a character. The first character entered will be your guess.")
                else if (guessedLetters.Contains(guess)) then
                    Console.WriteLine("You already guessed that letter!")
                else if (not(alphaCharReg.IsMatch(guess.ToString()))) then
                    Console.WriteLine("You can only guess English letters!")
                else
                    validGuess <- true
            guessedLetters.Add(guess)
            //Evaluate:
            let mutable didFindLetter : bool = false
            for i : int in 0..(fullWord.Length - 1) do
                if (fullWord.ToLower().[i] = guess) then
                    didFindLetter <- true
                    guessWord.Remove(i, 1) |> ignore
                    if (i = 0 || fullWord.[i - 1] = ' ') then
                        guessWord.Insert(i, guess.ToString().ToUpper()) |> ignore
                    else
                        guessWord.Insert(i, guess.ToString().ToLower()) |> ignore
            if (not(didFindLetter)) then
                livesLeft <- (livesLeft - 1)
            //Fresh start for printing UI.
            Console.Clear() |> ignore
            //Print status bar:
            Console.WriteLine("Lives Left: {0}.   Letters guessed: {1}\n", livesLeft, String.Join(", ", guessedLetters))
            //Print word:
            Console.WriteLine(guessWord.ToString())
            //Check for win/lose
            if (livesLeft = 0 || not(guessWord.ToString().Contains('_'))) then
                //you lose.
                if (livesLeft = 0) then
                    Console.WriteLine("Uh oh! Out of lives, you lose. The word/phrase was: \"{0}\" Play Again? (Y/N)", fullWord)
                else
                    Console.WriteLine("Congratulations! You Win! Play Again? (Y/N)")
                validInput <- false
                let mutable input : string = String.Empty
                while (not(validInput)) do
                    input <- Console.ReadLine()
                    if (affirmativeReg.IsMatch(input) || negativeReg.IsMatch(input)) then
                        validInput <- true
                    else
                        Console.WriteLine("Hey! That isn't a valid input, say \"Y\" or \"N\".")
                if (affirmativeReg.IsMatch(input)) then
                    livesLeft <- 6
                    guessedLetters.Clear()
                    guessWord.Clear() |> ignore
                    fullWord <- String.Empty
                    validInput <- false
                    gaming <- false
                    running <- true
                else if (negativeReg.IsMatch(input)) then
                    Console.WriteLine("Thanks for playing!")
                    Threading.Thread.Sleep(2000)
                    gaming <- false
                    running <- false
    0 // return an integer exit code